Run samples against Docker net-tools container

pfl opened the pull request:
Provide a script that runs the sample and starts a 'net-tools' container using the correct sample test Linux binary as counterpart. The Docker container executable will then return a success/not success value, whereby the script can deduce whether the sample completed successfully or not.

pfalcon reviewed (dismissed):
If the script introduced is supposed to perform heavy-weight operations, like running something in docker, can it please named in way which hints of that, instead of innocent name of "run-sample". If you don't have your own ideas, "run-sample-in-docker" would sound good to me.

nashif commented on samples/net/run-sample.sh:
```diff
@@ -0,0 +1,320 @@
+#!/bin/sh
```
Please put this script under scripts/, this is no place for scripts, for example scripts/net/...

pfl replied:
Yes, done.

pfl committed:
scripts: net: Add script for running sample tests

Add a script that sets up Docker networking and starts the net-tools
Docker container. If successful, run Zephyr with native_posix and
execute the appropriate net-tools container executable. The proper
net-tools executable and arguments is selected depending on the
basename of the sample directory. This script needs to be updated
if the net-tools Docker image is updated in some incompatible way.

The net-tools directory is assumed to exist at the same level as
the Zephyr base directory, but its location can be set using the
'-N' command line argument. Likewise, '-Z' sets the Zephyr top level
directory.

When stopping Zephyr, go through all child processes since running a
native posix or other variant of Zephyr may cause a hierarchy of
processes to be created.

Fixes: #19540

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

jukkar commented on samples/net/sockets/echo_client/prj.conf:
```diff
@@ -40,7 +40,7 @@ CONFIG_NET_SHELL=y
-CONFIG_NET_CONFIG_PEER_IPV6_ADDR="2001:db8::1"
-CONFIG_NET_CONFIG_MY_IPV4_ADDR="192.0.2.2"
-CONFIG_NET_CONFIG_PEER_IPV4_ADDR="192.0.2.1"
+CONFIG_NET_CONFIG_MY_IPV6_ADDR="2001:db8::1"
```
Ah, I missed the point when you asked the IP address numbering. I think we need to leave these like this atm so that qemu<->qemu communication works ok.
When compiling the echo-client, we can override the default addresses so can do like this: `cmake .... -DCONFIG_NET_CONFIG_MY_IPV6_ADDR="2001:db8::1" -DCONFIG_NET_CONFIG_PEER_IPV6_ADDR="2001:db8::2" -DCONFIG_NET_CONFIG_MY_IPV4_ADDR="192.0.2.1" -DCONFIG_NET_CONFIG_PEER_IPV4_ADDR="192.0.2.2"`

pfl replied:
Ok, let's keep the prj.conf and fix up the addresses assigned to the Docker image instead.

jukkar commented on scripts/net/run-sample-tests.sh:
```diff
@@ -0,0 +1,408 @@
+    if ! docker ps | grep "$name" > /dev/null
+    then
+	if ! docker run --hostname=$name --name=$name \
+             --ip=192.0.2.2 --ip6=2001:db8::2 \
```
What about setting the IP addresses somewhere at the beginning of the script, then it is easy to change them if needed.

pfl replied:
IP addresses (and other arguments) are now set depending on whether Zephyr is a(n echo) server or client.

jukkar commented on scripts/net/run-sample-tests.sh:
```diff
@@ -0,0 +1,408 @@
+{
+    rm -rf build && mkdir build && \
+	cmake -GNinja -DBOARD=native_posix -B build && \
+	ninja -C build
```
We could override the IP addresses here when compiling if needed (related to my earlier comment of echo-client prj.conf)

pfl replied:
Yes.

jukkar reviewed (changes requested):
Looks good. I think we need to think more about how to override the default IP numbering scheme.

nashif commented on scripts/net/run-sample-tests.sh:
```diff
@@ -0,0 +1,262 @@
+#!/bin/sh
```
commit message talks about echo_server, this however is not related to echo_server.

pfl replied:
Fixed.

nashif commented on CODEOWNERS:
```diff
@@ -328,6 +328,7 @@
 /scripts/gen_priv_stacks.py               @andrewboie @agross-oss @ioannisg
 /scripts/gen_syscall_header.py            @andrewboie
 /scripts/gen_syscalls.py                  @andrewboie
+/scripts/net/				  @jukkar @pfl
```
fix that spacing here.

pfl replied:
Tabs removed, spaces added.

pfl committed:
samples: net: echo_client: Add Kconfig option to run a number of times

Add Kconfig option NET_SAMPLE_SEND_ITERATIONS that sets the number of
times the Zephyr echo client sample sends its data. By default the
value is zero, which means indefinite, and demonstrates the same
behavior as before.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

jukkar commented on samples/net/sockets/echo_client/Kconfig:
```diff
@@ -46,4 +46,11 @@ config NET_SAMPLE_IFACE3_VLAN_TAG
 	  Set VLAN (virtual LAN) tag (id) that is used in the sample
```
The commit "samples: net: echo_client: Add Kconfig option to run a number of times" still says
```
For the echo_client, its prj.conf file defines
the value as 10, in which time the functionality is demonstrated
and the sample can be run in a meaningful way.
```
But as we removed the value from prj.conf this sentence can be removed.

pfl replied:
Done. And the license.

jukkar reviewed (commented):
Sanitychecker does not like MIT license, should we change this to Apache-2?

jukkar reviewed (approved):
LGTM

jukkar commented:
ping @nashif @pfalcon could you re-review the PR please

pfalcon commented:
I don't see my comment https://github.com/zephyrproject-rtos/zephyr/pull/19677#pullrequestreview-298668844 being addressed or replied to. However if there's no desire to make things easier to users (just today results of Zephyr UX study were presented at the TSC F2F meeting, and the results aren't stellar), let it be so.

pfl commented:
> I don't see my comment [#19677 (review)](https://github.com/zephyrproject-rtos/zephyr/pull/19677#pullrequestreview-298668844) being addressed or replied to. However if there's no desire to make things easier to users (just today results of Zephyr UX study were presented at the TSC F2F meeting, and the results aren't stellar), let it be so.

Here I tried to keep the naming generic/balanced. We might figure out something else than a Docker container to run the tests in; if so the script needs a major overhaul, but the "familiar" name would be kept. So I'm inclined to keep the name as not very docker specific as it is now.

pfalcon commented:
@pfl: Sounds good, thanks for reply.

jukkar commented:
ping @nashif can you recheck and give +1 if ok.

pfl commented:
ping @nashif were there any concerns still?

nashif commented:
few things:
- Can you squash all commits related to the script in one single commit? There is no value in getting all of this into the tree when the script is being submitted the first time. 
- There is no documentation or any guidance on how this script can or should be used.
- when I run it without any options, I am getting very cryptic output:

```
scripts/net/run-sample-tests.sh
$ZEPHYR_BASE /home/anashif/zephyrproject/zephyr
$NET_TOOLS_BASE /home/anashif/zephyrproject/zephyr/../net-tools
'zephyr' not supported
Sample 'zephyr' failed with return value '1'
Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Get http://%2Fvar%2Frun%2Fdocker.sock/v1.40/containers/json: dial unix /var/run/docker.sock: connect: permission denied
Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Get http://%2Fvar%2Frun%2Fdocker.sock/v1.40/networks: dial unix /var/run/docker.sock: connect: permission denied
```

- Don't you need to update west.yml to point to net-tools with docker image?

pfl commented:
> * Can you squash all commits related to the script in one single commit? There is no value in getting all of this into the tree when the script is being submitted the first time.

Done, there is now one commit for the script itself and another for the echo_client sample.

> * There is no documentation or any guidance on how this script can or should be used.

Added a README to help with that.
 
> * when I run it without any options, I am getting very cryptic output:

Fixed wording of cryptic output, now it mentions that it could not figure out which sample is wanted.

pfl commented:
ping @nashif ?

pfl commented:
ping @nashif ?

jukkar commented:
@nashif can you recheck this one?

nashif commented on scripts/net/README.txt:
```diff
@@ -0,0 +1,63 @@
+
+ * Install Docker
+ * Check out the net-tools project from github
+ * Change working directory to docker/ in the net-tools repository
```
is this step needed? I can't run the following command as specified from within the docker/ directory.

pfl replied:
Documentation fixed...

nashif commented on scripts/net/README.txt:
```diff
@@ -0,0 +1,63 @@
+ * Install Docker
+ * Check out the net-tools project from github
+ * Change working directory to docker/ in the net-tools repository
+ * Run './net-setup.sh --config docker.conf'
```
Maybe 'sudo ./net-setup.sh --config docker.conf'

jukkar replied:
The script will do sudo if needed.

nashif commented on scripts/net/README.txt:
```diff
@@ -0,0 +1,63 @@
+ * Change working directory to docker/ in the net-tools repository
+ * Run './net-setup.sh --config docker.conf'
+
+This creates a Docker image called 'net-tools' on your local machine.
```
Hmm, nothing happens here.. I get:

```
i9:net-tools((HEAD detached at 354fd05)): sudo ./net-setup.sh --config docker.conf
Using docker.conf configuration file.
Creating zeth
Error response from daemon: network with name net-tools0 already exists
bridge br- does not exist!
br-
```

pfl replied:
This is an issue with net-tools, fix is https://github.com/zephyrproject-rtos/net-tools/pull/40

nashif commented on scripts/net/README.txt:
```diff
@@ -0,0 +1,63 @@
+In essence, the following needs to be done:
+
+ * Install Docker
+ * Check out the net-tools project from github
```
net-tools is part of the west manifest, the current sha in west.yml is old and does not have docker/ directory, you need to update west.yml with the latest net-tools and point the user to it, i.e. `../tools/net-tools`. there is no need to checkout a new instance of it.

pfl replied:
Done.

nashif commented:
> -:322: WARNING:TYPO_SPELLING: 'childs' may be misspelled - perhaps 'children'?

maybe you want to address this as well :)

pfl committed:
west: Update required net-tools revision

Net-tools revision 4bff01084d225996e4aae84b98be5969e2f9f33d
is needed for running network sample test scripts.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl commented:
> maybe you want to address this as well :)

Yes, it is horrible with misspelled function names :-)

jukkar commented:
ping @nashif could you recheck?

pfl commented:
ping @nashif

jhedberg reviewed (approved):
Looks fine to me. Any chance @nashif that you could take another look before we merge?

pfl commented:
Ping @nashif ?
